File: cpp/endpoints/docking/import_docks.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_dock_msgs/srv/import_data.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("import_docks_client");
  rclcpp::Client<clearpath_dock_msgs::srv::ImportData>::SharedPtr client =
    node->create_client<clearpath_dock_msgs::srv::ImportData>("/" + namespace_ + "/" + std::string("docking/dock_manager/import"));

  auto request = std::make_shared<clearpath_dock_msgs::srv::ImportData::Request>();

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("import_docks_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("import_docks_client"), "/%s/docking/dock_manager/import service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("import_docks_client"), "Docks imported successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("import_docks_client"), "Failed to import docks!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("import_docks_client"), "Failed to call service /%s/docking/dock_manager/import", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}

File: python/endpoints/docking/import_docks.py
from clearpath_config.common.utils.yaml import read_yaml
import rclpy
from rclpy.node import Node
from clearpath_dock_msgs.srv import ImportData

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'ImportDocks'


class ImportDocks(Node):

    def __init__(self, namespace: str):
        super().__init__('import_docks_client')
        self._import_docks_client = self.create_client(ImportData, f'/{namespace}/docking/dock_manager/import')

        while not self._import_docks_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/docking/dock_manager/import service not available, waiting again...')

    def import_docks(self):
        req = ImportData.Request()
        self.future = self._import_docks_client.call_async(req)
        self.future.add_done_callback(self.import_docks_srv_response_cb)

    def import_docks_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Docks imported successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to import docks: {response.message}')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    import_docks_client = ImportDocks(namespace)
    import_docks_client.import_docks()

    rclpy.spin(import_docks_client)


if __name__ == '__main__':
    main()
